### PR TITLE
ACCUMULO-4065 Work around TExceptions being written back to clients i…

### PR DESCRIPTION
**Pull request by joshelser**
…n oneway methods.

In talking to @keith-turner on the previous pull request, we came to the conclusion that the real issue was that when a TException is thrown on by the (server-side) Processor on a oneway method, a message is written back to the client. This message is left on the wire, and will mess up other calls which expect their message to be on the wire (this was what the first pull request realized and tried to work around).

I noticed that RpcWrapper is actually rethrowing RTEs and Errors as TExceptions, in an attempt to work around some change in logic from 0.6.0 and 0.9.0. Since clients will never know that an Exception was thrown as a result of some invocation of a oneway Thrift call, we want to avoid wrapping and rethrowing any RTE/Error in the oneway call's path.

These changes alter RpcWrapper so that it does not wrap and rethrow RTEs and Errors when thrown from a oneway method invocation.

I verified that the patch works with some logging (thus removed), as well as hacking together a call to TabletClientService.fastHalt and verifying that no data is sent back to the client.

### CODE REVIEW

**Comment by keith-turner** [comment]
I think it would be useful to add some comments to the code pointing to ACCUMULO-4065 (and maybe THRIFT-1805 ACCUMULO-2950).  Looking at this code by itself someone would not know why it even exists.

File: server/base/src/main/java/org/apache/accumulo/server/util/RpcWrapper.java [hunk 1]
```diff
@@ -36,20 +42,51 @@
  * @since 1.6.1
  */
 public class RpcWrapper {
+  private static final Logger log = LoggerFactory.getLogger(RpcWrapper.class);
+
+  public static <T> T service(final T instance, @SuppressWarnings("rawtypes") final Map<String,ProcessFunction<T,? extends TBase>> processorView) {
+    // Get a handle on the isOnewayMethod and make it accessible
+    final Method isOnewayMethod;
+    try {
+      isOnewayMethod = ProcessFunction.class.getDeclaredMethod("isOneway");
```
Comment (keith-turner) [inline server/base/src/main/java/org/apache/accumulo/server/util/RpcWrapper.java]: ugh... looking at this code make me realize how brittle this new workaround and the existing work around are.   Brittle in the sense that we are expecting very specific Thrift internal behavior and method names, which could change in Thrift at any time.  I suppose we should look into pushing changes and/or test upstream that verify some sort of expected behavior.

**Comment by keith-turner** [comment]
Wonder what happens if a one-way throws a runtime exception and then the connection is put back into the pool.   My expectation is that a closing connection would be put back into the pool, which should be ok.   Whatever tries to use it later would hopefully get a ttransport exception, invalidate the connection, and retry.

File: server/base/src/main/java/org/apache/accumulo/server/util/RpcWrapper.java (hunk 1 above)
Comment (joshelser) [inline server/base/src/main/java/org/apache/accumulo/server/util/RpcWrapper.java]: Yeah, I didn't explicitly say, but I view this as a stop-gap for us to get a better fix into upstream.

**Comment by joshelser** [comment]
> Wonder what happens if a one-way throws a runtime exception and then the connection is put back into the pool

Now? Or with these changes? With these changes, the client will never see an error.

File: server/base/src/main/java/org/apache/accumulo/server/util/RpcWrapper.java (hunk 1 above)
Comment (keith-turner) [inline server/base/src/main/java/org/apache/accumulo/server/util/RpcWrapper.java]: ok  I made [a comment on THRIFT-1805](https://issues.apache.org/jira/browse/THRIFT-1805?focusedCommentId=15033849&page=com.atlassian.jira.plugin.system.issuetabpanels:comment-tabpanel#comment-15033849)

**Comment by keith-turner** [comment]
> Now? Or with these changes?

With these changes.
  
Based on reading @ctubbsii [comment on THRIFT-1805](https://issues.apache.org/jira/browse/THRIFT-1805?focusedCommentId=14042756&page=com.atlassian.jira.plugin.system.issuetabpanels:comment-tabpanel#comment-14042756), I am assuming that when you rethrow the runtime exception that the server side code will close the connection.   However whatever called the one-way may not be aware of this and would put the connection back into the pool.   I think it will all work out ok, but I was just curious what would happen in practice.

**Comment by joshelser** [comment]
>  I think it will all work out ok, but I was just curious what would happen in practice.

Ahh, I understand now. Your assessment seems accurate. Let me try to reproduce that. I should be able to with my little test setup.

**Comment by joshelser** [comment]
tl;dr Yes, the connection is unusable by the client, but the Accumulo code should retry that implicitly.

Expanded: After running my hacked-together `fastHalt` code to a TabletServer which didn't yet have its lock, I then made a call to `authenticate`. The call to `authenticate` fails with a `TTransportException -> IOException`. As such, `ServerClient` should implicitly catch and rerun the operation after a brief pause -- exactly what we want it to do.

**Comment by joshelser** [comment]
I need to add some unit tests for the new code added to RpcWrapper, but then I think this is ready to go.

WDYT?

File: server/base/src/main/java/org/apache/accumulo/server/util/RpcWrapper.java [hunk 2]
```diff
@@ -36,20 +42,51 @@
...
     InvocationHandler handler = new RpcServerInvocationHandler<T>(instance) {
       private final Logger log = LoggerFactory.getLogger(instance.getClass());
 
       @Override
       public Object invoke(Object obj, Method method, Object[] args) throws Throwable {
+        // e.g. ThriftClientHandler.flush(TInfo, TCredentials, ...)
         try {
           return super.invoke(obj, method, args);
         } catch (RuntimeException e) {
+          if (onewayMethods.contains(method.getName())) {
```
Comment (keith-turner) [inline server/base/src/main/java/org/apache/accumulo/server/util/RpcWrapper.java]: Does entire method signature need to be considered?

File: server/base/src/main/java/org/apache/accumulo/server/util/RpcWrapper.java [hunk 3]
```diff
@@ -36,20 +42,51 @@
...
+    // Get a handle on the isOnewayMethod and make it accessible
+    final Method isOnewayMethod;
+    try {
+      isOnewayMethod = ProcessFunction.class.getDeclaredMethod("isOneway");
+    } catch (NoSuchMethodException e) {
+      throw new RuntimeException("Could not access isOneway method", e);
+    } catch (SecurityException e) {
+      throw new RuntimeException("Could not access isOneway method", e);
+    }
+    isOnewayMethod.setAccessible(true);
```
Comment (keith-turner) [inline server/base/src/main/java/org/apache/accumulo/server/util/RpcWrapper.java]: Should this be set back to false?

**Comment by keith-turner** [comment]
> WDYT?

I just made two more comments.  Other than those I am finished looking at the changes.

File: server/base/src/main/java/org/apache/accumulo/server/util/RpcWrapper.java (hunk 3 above)
Comment (joshelser) [inline server/base/src/main/java/org/apache/accumulo/server/util/RpcWrapper.java]: No, we need it to be accessible to invoke the method later, I believe. I will double check the API -- I don't use reflection enough to remember this all for certain.

File: server/base/src/main/java/org/apache/accumulo/server/util/RpcWrapper.java (hunk 2 above)
Comment (joshelser) [inline server/base/src/main/java/org/apache/accumulo/server/util/RpcWrapper.java]: I believe that Thrift disallows multiple methods with the same name in one service, but I will dbl check.

Comment (joshelser) [inline server/base/src/main/java/org/apache/accumulo/server/util/RpcWrapper.java]: > I believe that Thrift disallows multiple methods with the same name in one service

Yes, this is the case. Each Thrift service (e.g. master, client, tabletserver) will have its own instance of RpcWrapper underneath the `Processor`. Thus, the methods contained are unique to a service definition whose method names are guaranteed to be unique.

File: server/base/src/main/java/org/apache/accumulo/server/util/RpcWrapper.java (hunk 3 above)
Comment (joshelser) [inline server/base/src/main/java/org/apache/accumulo/server/util/RpcWrapper.java]: > No, we need it to be accessible to invoke the method later, I believe

Right, `setAccessible` makes the `protected isOneway` method available for us to bypass Java security and actually invoke the method. We need to keep this instance of the `Method` accessible.

File: server/base/src/main/java/org/apache/accumulo/server/util/RpcWrapper.java [hunk 4]
```diff
@@ -36,20 +42,51 @@
...
+    final Set<String> onewayMethods = new HashSet<String>();
+    for (@SuppressWarnings("rawtypes") Entry<String,ProcessFunction<T,? extends TBase>> entry : processorView.entrySet()) {
+      try {
+        if ((Boolean) isOnewayMethod.invoke(entry.getValue())) {
+          onewayMethods.add(entry.getKey());
+        }
+      } catch (Exception e) {
+        throw new RuntimeException(e);
+      }
+    }
```
Comment (keith-turner) [inline server/base/src/main/java/org/apache/accumulo/server/util/RpcWrapper.java]: can we restore isOneWayMethod's accessible settings here?

File: server/base/src/main/java/org/apache/accumulo/server/util/RpcWrapper.java (hunk 2 above)
Comment (keith-turner) [inline server/base/src/main/java/org/apache/accumulo/server/util/RpcWrapper.java]: ok, in that case could add a sanity check to the code above that adds method names to the set.   When add is called on the set it should always return true.   If add returns false, then throw a new WTF exception

File: server/base/src/main/java/org/apache/accumulo/server/util/RpcWrapper.java [hunk 5]
```diff
@@ -36,20 +42,51 @@
...
+    isOnewayMethod.setAccessible(true);
+
+    final Set<String> onewayMethods = new HashSet<String>();
+    for (@SuppressWarnings("rawtypes") Entry<String,ProcessFunction<T,? extends TBase>> entry : processorView.entrySet()) {
+      try {
+        if ((Boolean) isOnewayMethod.invoke(entry.getValue())) {
+          onewayMethods.add(entry.getKey());
+        }
+      } catch (Exception e) {
```
Comment (keith-turner) [inline server/base/src/main/java/org/apache/accumulo/server/util/RpcWrapper.java]: could add following above this.. this avoid wrapping runtime with runtime... 
```java
} catch (RuntimeException e) { 
   throw e
```

File: server/base/src/main/java/org/apache/accumulo/server/util/RpcWrapper.java (hunk 4 above)
Comment (joshelser) [inline server/base/src/main/java/org/apache/accumulo/server/util/RpcWrapper.java]: Cool, let me dbl check. Eclipse didnt' want to pull down the sources for some of the sun classes, but the implementation of getMethod certainly looked like it made a copy.

**Comment by joshelser** [comment]
Ok, lots of good stuff contained in 22751b7a.

* Lots of comments
* Lots of unit tests for RpcWrapper (includes some moving around in RpcWrapper to ease testing).
* Avoid rewrapping RTE (Keith comment)
* Restore `isAccessbile` on the Method (even though it's probably unnecessary -- good insulation)

Pending a `mvn verify -Psunny`. I'll squash and merge this into 1.6, 1.7 and master.

File: server/tserver/src/main/java/org/apache/accumulo/tserver/TabletServer.java [hunk 6]
```diff
@@ -3159,7 +3159,8 @@ private void returnMasterConnection(MasterClientService.Client client) {
 
   private HostAndPort startTabletClientService() throws UnknownHostException {
     // start listening for client connection last
-    Iface tch = RpcWrapper.service(new ThriftClientHandler());
+    ThriftClientHandler handler = new ThriftClientHandler();
+    Iface tch = RpcWrapper.service(handler, new Processor<Iface>(handler).getProcessMapView());
```
Comment (ctubbsii) [inline server/tserver/src/main/java/org/apache/accumulo/tserver/TabletServer.java]: For all of these calls, I think it's better to pass the Processor, and call the `getProcessMapView()` in the method. It makes it a lot more readable, especially with all the generics.

File: proxy/src/main/java/org/apache/accumulo/proxy/Proxy.java [hunk 7]
```diff
@@ -135,7 +136,8 @@ public static TServer createProxyServer(Class<?> api, Class<?> implementor, fina
     @SuppressWarnings("unchecked")
```
Comment (ctubbsii) [inline proxy/src/main/java/org/apache/accumulo/proxy/Proxy.java]: Before, this suppression was fine to do here, but I think the cast to `Class<TProcessor>` should be moved up to the assignment of `proxyProcClass` instead. That way, the cast here for the `Constructor` is unnecessary, and makes it easier to follow to the next line when the `Constructor` is instantiated with the `RpcWrapper`.

File: proxy/src/main/java/org/apache/accumulo/proxy/Proxy.java [hunk 8]
```diff
@@ -135,7 +136,8 @@ public static TServer createProxyServer(Class<?> api, Class<?> implementor, fina
...
     Constructor<? extends TProcessor> proxyProcConstructor = (Constructor<? extends TProcessor>) proxyProcClass.getConstructor(proxyIfaceClass);
 
-    final TProcessor processor = proxyProcConstructor.newInstance(RpcWrapper.service(impl));
+    @SuppressWarnings({"rawtypes", "unchecked"})
```
Comment (ctubbsii) [inline proxy/src/main/java/org/apache/accumulo/proxy/Proxy.java]: Neither rawtypes, nor unchecked are necessary to suppress here if we cast to `TBaseProcessor<?>`.

**Comment by joshelser** [comment]
So it doesn't look like I just ignored Christopher -- we were talking in IRC about his recommendations. He opened up joshelser/accumulo#1, but said he can just merge it tmrw morning. I plan to leave that to him.
